drivers/include/periph/wdt.h

```diff
@@ -0,0 +1,169 @@
+ *    wdt_start();
+ *
+ *    // MIN_TIME < VALID_TIME < MAX_TIME
+ *    while (1) {
+ *        xtimer_usleep(VALID_TIME);
+ *        wdt_kick()
+ *    }
+ *
+ * |---------------------MAX_TIME-----------------------|
+ * |---------MIN_TIME-------|-----------WINDOW----------|
+ *
+ * |------------VALID_TIME-----------|
+ *                                   ^
+ *                                   |
+ *                              wdt_kick()
+ * 
+ * WDT callback
+ * ============
+ *
+ * Before reboot WDT can trigger Early Wakeup Interrupt and execute a callback
+ * to perform specific safety operations of data logging before the actual reboot.
+ *
+ *    wdt_setup_callback(MIN_TIME, MAX_TIME, wdt_cb, arg);
+ *
+ * To include this feature, (If your platform supports it) in your app Makefile add:
+ *
+ *    MODULE += periph_wdt_cb
+ *
+ *
+ * @{
+ *
+ * @file        wdt.h
+ * @brief       watchdog peripheral interface definitions
+ *
+ * @author      Thomas Geithner <thomas.geithner@dai-labor.de>
+ *              Francisco Molina <francois-xavier.molina@inria.fr>
+ *              Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ */
+
+#ifndef PERIPH_WDT_H
+#define PERIPH_WDT_H
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifdef MODULE_PERIPH_WDT_CB
+/**
+ * @brief Signature for the watchdog early warning callback
+ *
+ * @param[in] arg           optional argument which is passed to the
+ *                          callback
+ */
+typedef void(*wdt_cb_t)(void *arg);
+#endif
+
+/**
+ * @brief    Start watchdog timer
+ */
+void wdt_start(void);
+
+/**
+ * @brief    Stop watchdog timer
+ */
+void wdt_stop(void);
+
+/**
+ * @brief    Reset the watchdog timer counter, delay system reset
+ */
+void wdt_kick(void);
+
+/**
+ * @brief    Set up the wdt timer, only use max_time if normal operation
+ *           set min_time and max_time for windowed timer.
+ *
+ * @param[in] min_time       lower bound for windowed watchdog in ms, has to be 0
+ *                           for normal mode operation
+ * @param[in] max_time       upper bound for windowed watchdog in ms, time before
+ *                           reset for normal watchdog
+ */
+void wdt_setup_reboot(uint32_t min_time, uint32_t max_time);
+
+#ifdef MODULE_PERIPH_WDT_CB
+/**
+ * @brief    Set up the wdt timer, only use max_time if normal operation
+ *           set min_time and max_time for windowed timer.
+ *
+ * @param[in] min_time       lower bound for windowed watchdog in ms, has to be 0
+ *                           for normal mode operation
+ * @param[in] max_time       upper bound for windowed watchdog in ms, time before
+ *                           reset for normal watchdog
+ * @param[in] wdt_cb         wdt callback, can be NULL
+ * @param[in] arg            optional argument which is passed to the
+ *                           callback, can be NULL
+ */
+void wdt_setup_callback(uint32_t min_time, uint32_t max_time, wdt_cb_t wdt_cb, void* arg);
```

Comment:
I think this can be misleading: The watchdog will still trigger a reboot and behave just like with `wdt_setup_reboot`, the only difference being that `wdt_cb` is executed before the watchdog timer expires.

Having two functions for it make it seem like `wdt_setup_callback` would cause the callback to be called *instead* of the reboot, not in addition to it.

Imho one `wdt_setup(uint32_t min_time, uint32_t max_time, wdt_cb_t wdt_cb, void* arg)` should be enough. `wdt_cb` can be `NULL` if it's not needed.

Also it would be good if the function can return an error code if an invalid configuration is requested.

Reply:
> I think this can be misleading: The watchdog will still trigger a reboot and behave just like with wdt_setup_reboot, the only difference being that wdt_cb is executed before the watchdog timer expires.

Maybe call it ```wdt_setup_reset_with_warning()```?

BTW, we need an additional define for the time difference between warning callback and actual reset.
 
> Having two functions for it make it seem like `wdt_setup_callback` would cause the callback to be called _instead_ of the reboot, not in addition to it.

If this is one function, the functionality cannot be guarded properly. When supported, all callback handling code must be compiled in, even if unused.
In practice, ```wdt_setup_reset_with_warning()``` will probably look like

```
wdt_setup_reset_with_warning() {
 wdt_setup_reset();
// set callback
}
```

> Also it would be good if the function can return an error code if an invalid configuration is requested.

Would be good for what? Can we get away with asserts and log messages?

```
if (wdt_setup_reset() == EINVAL) {
  // now what? set a flag disabling wdt_kick()?
  // let the user know?
}
```

Reply:
I'm fine with `wdt_setup_reset_with_warning()` if that means that the compiler can LTO away those pesky 8 bytes if not used. 

Log messages on the other hand tend to be larger than 8 byte (yea, ROM hurts a little less than RAM, but still), but I guess we are good when we enable them when `DEVELHELP` is set.

Reply:
> I think this can be misleading: The watchdog will still trigger a reboot and behave just like with wdt_setup_reboot, the only difference being that wdt_cb is executed before the watchdog timer expires.

> BTW, we need an additional define for the time difference between warning callback and actual 
reset.

For sam21 its configurable in normal mode but matches the start of the window in windowed mode (which seems useless to me), for nrf52 its a different configurable value and for stm32 its fixed at when the downcounter hits 0x40. A define would work for stm32, nrf52 but would not work in the case of windowed operation for sam21, but proper DEBUG messages could state this behavior. 

> Having two functions for it make it seem like wdt_setup_callback would cause the callback to be called instead of the reboot, not in addition to it.

Splitting into two functions makes sense to me since we really have very different behaviors for the callback according to each platform, while normal operation is pretty uniform.

> Can we get away with asserts and log messages?

+1